backbone/src/main/java/org/researchstack/backbone/ui/task/TaskViewModel.kt

```diff
@@ -280,4 +278,47 @@ internal class TaskViewModel(val context: Application, intent: Intent) : Android
         stack.clear()
         editing = false
     }
+
+    fun checkForSaveDialog(stepResult: StepResult<*>?) {
```

Comment:
This is a complicated method, very hard to read to me. Don't you feel the same?

How about a less fancy/more "open" version. For example, a variation of:
```
    fun checkForSaveDialog(stepResult: StepResult<*>?) {
        when {
            isSavedDialogAppeared.not() && checkIfAnswersAreTheSame(stepResult).not() && checkIfCurrentStepIsBranchDecisionStep() -> {
                isSavedDialogAppeared = true
                showSaveEditDialog.postValue(true)
            }
            currentStep!!.isOptional && checkIfTheNewAnswerIsSkipWhileThePreviousIsNot(stepResult) -> {
                isSavedDialogAppeared = true
                showSaveEditDialog.postValue(true)
            }
            else -> {
                nextStep()
            }
        }
    }

```

Does that read better to you?